File: rpi_hailo_python/src/dataset_writer.py
import os
import cv2
import time
import uuid
import random
import sqlite3
import logging
from datetime import datetime
from .utils import ensure_directories

logger = logging.getLogger("DatasetWriter")

class DatasetWriter:
    def __init__(self, config):
        self.config = config['storage']
        self.base_path = self.config['base_path']
        self.train_split = self.config.get('train_split', 0.8)
        
        self.images_dir = self.config['images_dir']
        self.labels_dir = self.config['labels_dir']
        self.db_path = os.path.join(self.base_path, self.config.get('database_path', 'datacollector.db'))
        
        self.setup_directories()
        self.setup_database()
        
    def setup_directories(self):
        # Create YOLO structure
        # dataset/images/train, dataset/images/val
        # dataset/labels/train, dataset/labels/val
        subdirs = [
            os.path.join(self.images_dir, 'train'),
            os.path.join(self.images_dir, 'val'),
            os.path.join(self.labels_dir, 'train'),
            os.path.join(self.labels_dir, 'val')
        ]
        ensure_directories(self.base_path, subdirs)
        logger.info(f"Dataset directories initialized at {self.base_path}")

    def setup_database(self):
        conn = sqlite3.connect(self.db_path)
        c = conn.cursor()
        c.execute('''
            CREATE TABLE IF NOT EXISTS frames (
                id TEXT PRIMARY KEY,
                camera_id TEXT,
                timestamp REAL,
                split TEXT,
                image_path TEXT,
                label_path TEXT,
                objects_count INTEGER,
                classes TEXT
            )
        ''')
        conn.commit()
        conn.close()

    def save_sample(self, frame, camera_id, annotations, classes_detected):
        """
        Save a frame and its annotations.
        frame: numpy array
        camera_id: str
        annotations: list of YOLO format strings
        classes_detected: list of class names/ids found
        """
        if not annotations and not self.config.get('save_empty', False):
            return

        timestamp = time.time()
        frame_id = f"{camera_id}_{int(timestamp * 1000)}_{str(uuid.uuid4())[:8]}"
        
        # Determine split
        split = 'train' if random.random() < self.train_split else 'val'
        
        # Paths
        img_filename = f"{frame_id}.jpg"
        lbl_filename = f"{frame_id}.txt"
        
        img_rel_path = os.path.join(self.images_dir, split, img_filename)
        lbl_rel_path = os.path.join(self.labels_dir, split, lbl_filename)
        
        img_full_path = os.path.join(self.base_path, img_rel_path)
        lbl_full_path = os.path.join(self.base_path, lbl_rel_path)
        
        # Save Image
        cv2.imwrite(img_full_path, frame)
        
        # Save Label
        with open(lbl_full_path, 'w') as f:
            f.write('\n'.join(annotations))
            
        # Log to DB
        self.log_to_db(frame_id, camera_id, timestamp, split, img_rel_path, lbl_rel_path, len(annotations), str(classes_detected))
        
        logger.debug(f"Saved sample {frame_id} to {split}")

    def log_to_db(self, frame_id, camera_id, timestamp, split, img_path, lbl_path, count, classes):
        try:
            conn = sqlite3.connect(self.db_path)
            c = conn.cursor()
            c.execute('''
                INSERT INTO frames (id, camera_id, timestamp, split, image_path, label_path, objects_count, classes)
                VALUES (?, ?, ?, ?, ?, ?, ?, ?)
            ''', (frame_id, camera_id, timestamp, split, img_path, lbl_path, count, classes))
            conn.commit()
            conn.close()
        except Exception as e:
            logger.error(f"Failed to log to DB: {e}")

File: rpi_hailo_cpp/src/DatasetWriter.cpp
#include "DatasetWriter.hpp"
#include <filesystem>
#include <fstream>
#include <iostream>
#include <chrono>
#include <random>

namespace fs = std::filesystem;

DatasetWriter::DatasetWriter(const StorageConfig& cfg) : config(cfg), db(nullptr), insert_stmt(nullptr) {
    setupDirectories();
    setupDatabase();
    prepareStatement();
}

DatasetWriter::~DatasetWriter() {
    if (insert_stmt) sqlite3_finalize(insert_stmt);
    if (db) sqlite3_close(db);
}

void DatasetWriter::setupDirectories() {
    std::vector<std::string> subs = {"images/train", "images/val", "labels/train", "labels/val"};
    for (const auto& sub : subs) {
        fs::create_directories(fs::path(config.base_path) / sub);
    }
}

void DatasetWriter::setupDatabase() {
    std::string db_path = (fs::path(config.base_path) / config.database_path).string();
    int rc = sqlite3_open(db_path.c_str(), &db);
    if (rc) {
        std::cerr << "Can't open database: " << sqlite3_errmsg(db) << std::endl;
        return;
    }
    
    // Create table if not exists
    const char* sql = "CREATE TABLE IF NOT EXISTS frames (" \
                      "id TEXT PRIMARY KEY, camera_id TEXT, " \
                      "timestamp REAL, split TEXT, image_path TEXT);";
    char* errMsg = 0;
    rc = sqlite3_exec(db, sql, 0, 0, &errMsg);
    if (rc != SQLITE_OK) {
        std::cerr << "SQL error: " << errMsg << std::endl;
        sqlite3_free(errMsg);
    }
    
    // Enable WAL mode for better concurrency
    sqlite3_exec(db, "PRAGMA journal_mode=WAL;", 0, 0, 0);
}

void DatasetWriter::prepareStatement() {
    if (!db) return;
    const char* sql = "INSERT INTO frames (id, camera_id, split, image_path) VALUES (?, ?, ?, ?);";
    int rc = sqlite3_prepare_v2(db, sql, -1, &insert_stmt, 0);
    if (rc != SQLITE_OK) {
        std::cerr << "Failed to prepare statement: " << sqlite3_errmsg(db) << std::endl;
    }
}

void DatasetWriter::saveSample(const cv::Mat& frame, 
                             const std::string& camera_id, 
                             const std::vector<std::string>& yolo_lines,
                             const std::vector<std::string>& classes) {
    // 1. Generate ID
    auto now = std::chrono::system_clock::now();
    auto timestamp = std::chrono::duration_cast<std::chrono::milliseconds>(now.time_since_epoch()).count();
    std::string frame_id = camera_id + "_" + std::to_string(timestamp);

    // 2. Determine split
    static std::random_device rd;
    static std::mt19937 gen(rd());
    static std::uniform_real_distribution<> dis(0.0, 1.0);
    std::string split = (dis(gen) < config.train_split) ? "train" : "val";

    // 3. Paths
    std::string img_name = frame_id + ".jpg";
    std::string lbl_name = frame_id + ".txt";
    
    fs::path img_path = fs::path(config.base_path) / "images" / split / img_name;
    fs::path lbl_path = fs::path(config.base_path) / "labels" / split / lbl_name;

    // 4. Save
    cv::imwrite(img_path.string(), frame);
    
    std::ofstream out(lbl_path);
    for (const auto& line : yolo_lines) {
        out << line << "\n";
    }
    out.close();

    // 5. Log
    logToDb(frame_id, camera_id, split, img_path.string());
    std::cout << "Saved " << frame_id << " to " << split << std::endl;
}

void DatasetWriter::logToDb(const std::string& id, const std::string& cam_id, 
                          const std::string& split, const std::string& img_path) {
    if (!db || !insert_stmt) return;

    // Bind parameters
    // Index starts at 1
    sqlite3_bind_text(insert_stmt, 1, id.c_str(), -1, SQLITE_STATIC);
    sqlite3_bind_text(insert_stmt, 2, cam_id.c_str(), -1, SQLITE_STATIC);
    sqlite3_bind_text(insert_stmt, 3, split.c_str(), -1, SQLITE_STATIC);
    sqlite3_bind_text(insert_stmt, 4, img_path.c_str(), -1, SQLITE_STATIC);

    // Execute
    if (sqlite3_step(insert_stmt) != SQLITE_DONE) {
        std::cerr << "DB Insert Failed: " << sqlite3_errmsg(db) << std::endl;
    }

    // Reset for next use
    sqlite3_reset(insert_stmt);
    sqlite3_clear_bindings(insert_stmt);
}
